devicetree: Add a zephyr/zephyr.dts debugging aid that has the final devicetree

ulfalizer opened the pull request:
(Screenshots still show `merged.dts`, but I renamed it to `zephyr.dts`.)

Adds a new `zephyr/zephyr.dts` output that holds the final devicetree, as a debugging aid. `zephyr.dts` preserves labels in phandle and path references too.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/22272.

Also redraws a diagram in the devicetree section of the manual to be accurate after the old devicetree scripts were removed.

Before:
![io-before](https://user-images.githubusercontent.com/154768/73343799-7cc94100-4281-11ea-9d89-0ac4d2fdb7f9.png)

After:
![io-after](https://user-images.githubusercontent.com/154768/73343823-85217c00-4281-11ea-8e97-17ae7ed9601d.png)

`zephyr.dts` is included in the build overview diagram now too:

![Skärmbild från 2020-01-29 10-26-43](https://user-images.githubusercontent.com/154768/73344041-eba69a00-4281-11ea-8b69-27cebbb90143.png)

Commits:

```
dts: edtlib: Add an EDT.dts_source attribute that holds the source code

EDT.dts_source is the source code of the loaded .dts file, after merging
nodes.
```

```
dts: gen_defines.py: Save final devicetree to zephyr/zephyr.dts

Add a '--dts-out <file>' flag to gen_defines.py that saves the final
devicetree to <file>, in DTS format, using the new EDT.dts_source
attribute.

Handy to have available as a debugging aid. Unused otherwise.
```

```
doc: devicetree: Update outdated diagram and mention zephyr.dts

zephyr_dt_inputs_outputs.svg still shows the output from dtc being used,
but dtc is unused (except for finding warnings/errors) after the old
devicetree scripts were removed in commit c8c35f76ab ("scripts: dts:
Remove deprecated extract_dts_includes.py script").

Update zephyr_dt_inputs_outputs.svg to show how things are done now.
Also include the new zephyr.dts debugging aid in it. Tweak the
formatting a bit too.

Add zephyr.dts to the diagram in the build overview section too, and
mention zephyr.dts in the text of the devicetree and build overview
pages.

Remove zephyr_dt_inputs_outputs.png and use zephyr_dt_inputs_outputs.svg
directly. Many other places the documentation include SVGs directly, and
there haven't been any complaints, so it probably works fine. The .png
and .svg versions had also drifted out of sync.

Piggyback a link from the devicetree page to the build overview page, to
make it easier to discover.

(I used draw.io to update the diagrams.)
```

ulfalizer committed:
dts: edtlib: Add an EDT.dts_source attribute that holds the source code

EDT.dts_source is the source code of the loaded .dts file, after merging
nodes.

Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

ulfalizer committed:
dts: gen_defines.py: Save final devicetree to zephyr/zephyr.dts

Add a '--dts-out <file>' flag to gen_defines.py that saves the final
devicetree to <file>, in DTS format, using the new EDT.dts_source
attribute.

Handy to have available as a debugging aid. Unused otherwise.

Also write a dummy <BOARD>.dts_compiled file that tells people to look
at zephyr.dts, for people that might be used to that file. It was
removed in commit 8d8b06978c ("dts: Remove generation of
<BOARD>.dts_compiled").

Fixes: #22272

Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

ulfalizer committed:
doc: devicetree: Update outdated diagram and mention zephyr.dts

zephyr_dt_inputs_outputs.svg still shows the output from dtc being used,
but dtc is unused (except for finding warnings/errors) after the old
devicetree scripts were removed in commit c8c35f76ab ("scripts: dts:
Remove deprecated extract_dts_includes.py script").

Update zephyr_dt_inputs_outputs.svg to show how things are done now.
Also include the new zephyr.dts debugging aid in it. Tweak the
formatting a bit too.

Add zephyr.dts to the diagram in the build overview section too, and
mention zephyr.dts in the text of the devicetree and build overview
pages.

Remove zephyr_dt_inputs_outputs.png and use zephyr_dt_inputs_outputs.svg
directly. Many other places the documentation include SVGs directly, and
there haven't been any complaints, so it probably works fine. The .png
and .svg versions had also drifted out of sync.

Piggyback a link from the devicetree page to the build overview page, to
make it easier to discover.

(I used draw.io to update the diagrams.)

Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

ulfalizer commented:
Skipped the board name in `zephyr/merged.dts` since the board is already known. Nice to have a simple, consistent filename.

pabigot reviewed (changes requested):
I could bikeshed the name (final.dts, application.dts, image.dts) but other than not really liking merged.dts I can't immediately identify a clearly superior solution.  Except maybe zephyr.dts for consistency with zephyr.elf and zephyr.map and other final outputs that end up in that directory.

Yeah, I'd like zephyr.dts.

Are we guaranteed that the devicetree source produced by this library can be processed by dtc?  Because I can see doing that.

ulfalizer commented:
> I could bikeshed the name (final.dts, application.dts, image.dts) but other than not really liking merged.dts I can't immediately identify a clearly superior solution. Except maybe zephyr.dts for consistency with zephyr.elf and zephyr.map and other final outputs that end up in that directory.
> 
> Yeah, I'd like zephyr.dts.
> 

`merged.dts` seems more specific to me, so I like it better.

I'd probably go "hmm... wonder if I can see the final devicetree somewhere... ohh... merged.dts must be it." With `zephyr.dts`, I'd go "I wonder if this is some default devicetree thingy that gets included by everything" instead.

So `merged.dts` might be better from a psychology perspective.

> Are we guaranteed that the devicetree source produced by this library can be processed by dtc? Because I can see doing that.

Should be. Otherwise there's a bug, and it should be pretty stable, because it gets tested a lot indirectly in `testdtlib.py`.

ulfalizer commented:
`final.dts` doesn't seem too bad either though. Not against going with that (except I'd have to wrestle draw.io for a few minutes again).

pabigot commented:
> `merged.dts` seems more specific to me, so I like it better.
> 
> I'd probably go "hmm... wonder if I can see the final devicetree somewhere... ohh... merged.dts must be it." With `zephyr.dts`, I'd go "I wonder if this is some default devicetree thingy that gets included by everything" instead.

I don't think that makes sense from a developer perspective.  We tend to look for `zephyr.{bin,elf,hex,lst,map,stat}` in that directory as being associated with the final image product.  `merged.dts` doesn't suggest it's associated with the final product; it could be an intermediate file.  If I want a dt blob that corresponds to the image needs for some reason, I'd look for `zephyr.dts` to produce it.

@mbolivar ?

ulfalizer commented:
> > `merged.dts` seems more specific to me, so I like it better.
> > I'd probably go "hmm... wonder if I can see the final devicetree somewhere... ohh... merged.dts must be it." With `zephyr.dts`, I'd go "I wonder if this is some default devicetree thingy that gets included by everything" instead.
> 
> I don't think that makes sense from a developer perspective. We tend to look for `zephyr.{bin,elf,hex,lst,map,stat}` in that directory as being associated with the final image product. `merged.dts` doesn't suggest it's associated with the final product; it could be an intermediate file. If I want a dt blob that corresponds to the image needs for some reason, I'd look for `zephyr.dts` to produce it.

Ah, yeah, looking at it that way it does make some sense. Still would've been confused though, because I never thought if it that way (anything with zephyr in the name = final output).

Then I won't fight it at least. I'll rename it if it helps the PR go in faster.

mbolivar commented:
I think I'd prefer either not renaming it at all (BOARD.dts_compiled) or using zephyr.dts.

Actually, could you create BOARD.dts_compiled for zephyr 2.2 in addition to zephyr.dts, even if its contents just say "go look at zephyr.dts instead", similarly to how you've kept the old header files around for now?

IMO developers will be looking for this file, and will be confused by its disappearance. Having a shim for one release will help.

ulfalizer commented:
Renamed the file to `zephyr.dts`.

> Actually, could you create BOARD.dts_compiled for zephyr 2.2 in addition to zephyr.dts, even if its contents just say "go look at zephyr.dts instead", similarly to how you've kept the old header files around for now?
> 
> IMO developers will be looking for this file, and will be confused by its disappearance. Having a shim for one release will help.

Added. `cmake/dts.cmake` now writes a dummy `<BOARD>.dts_compiled`.

SebastianBoe commented on cmake/dts.cmake:
```diff
@@ -209,6 +210,11 @@ if(SUPPORTS_DTS)
+  # equivalent.
+  file(WRITE ${PROJECT_BINARY_DIR}/${BOARD}.dts_compiled
+       "See zephyr.dts for the final merged devicetree.")
+
```
Let's add a comment about removing this after the next Zephyr release.

We don't want this noise here forever.

ulfalizer replied:
Added "will be removed in Zephyr 2.3."

pabigot commented on cmake/dts.cmake:
```diff
@@ -198,6 +198,7 @@ if(SUPPORTS_DTS)
   --bindings-dirs ${DTS_ROOT_BINDINGS}
   --conf-out ${DEVICETREE_CONF}
   --header-out ${DEVICETREE_UNFIXED_H}
+  --dts-out ${PROJECT_BINARY_DIR}/merged.dts # As a debugging aid
```
This is still generating `merged.dts` rather than `zephyr.dts`.

ulfalizer replied:
Oops... search fail. Fixed.

Had missed a spot in the docs too.

galak commented:
Generally LGTM, once you guys decide on the details.

mbolivar commented:
From the docs commit message:

> Many other places the documentation include SVGs directly, and there haven't been any complaints, so it probably works fine.

FWIW, I created a PNG specifically because the SVG was not rendering properly in Firefox on Linux. Some of the text was either getting truncated or spilling out of the bounding box of the block that contained it. I left the SVG in there to make future edits easier.

Have you tested this on a few different browsers?

mbolivar commented:
> Some of the text was either getting truncated or spilling out of the bounding box of the block that contained it.

Found the comment with the details: https://github.com/zephyrproject-rtos/zephyr/pull/19630#issuecomment-538582774

ulfalizer commented:
> FWIW, I created a PNG specifically because the SVG was not rendering properly in Firefox on Linux. Some of the text was either getting truncated or spilling out of the bounding box of the block that contained it. I left the SVG in there to make future edits easier.
> 
> Have you tested this on a few different browsers?

That one seems to render fine in Firefox at least:

![Skärmbild från 2020-01-29 17-12-58](https://user-images.githubusercontent.com/154768/73374615-cbdf9800-42ba-11ea-883c-6db158d37103.png)

Looks good in Chromium too.

Wonder if something might be wonky with dimensions if other ones are broken.

mbolivar reviewed (approved):
I double-checked the documentation in my browser and the SVG looks good there too. Looks good to me, thanks!
